### PR TITLE
memdb: use a bit rand generator

### PR DESCRIPTION
**Pull request by tiancaiamao**
go test -bench shows no improvement, in a single goroutine environment
not sure it really work @coocood 

https://github.com/pingcap/tidb/issues/1788

### CODE REVIEW

**Comment by coocood** [comment]
We can just benchmark concurrently Put on multiple memDB.

**Comment by tiancaiamao** [comment]
under 5 concurrent goroutines get **3%** improvement
```
BenchmarkConcurrentPut-4   	  500000	      5182 ns/op
BenchmarkConcurrentPut-4   	  500000	      5012 ns/op
```

under 50 concurrent goroutines get **13%** improvement

```
BenchmarkConcurrentPut-4   	   50000	     33310 ns/op
BenchmarkConcurrentPut-4   	  100000	     29356 ns/op
```

under 100 concurrent goroutines get **40%** improvement
```
BenchmarkConcurrentPut-4   	   30000	     68582 ns/op
BenchmarkConcurrentPut-4   	   30000	     48185 ns/op
```

@coocood

File: leveldb/memdb/memdb.go [hunk 1]
```diff
@@ -38,9 +38,7 @@ func (r *lockedSource) Seed(seed int64) {
 var (
 	ErrNotFound     = errors.ErrNotFound
 	ErrIterReleased = errors.New("leveldb/memdb: iterator released")
-	rnd             = rand.New(&lockedSource{
-		src: rand.NewSource(0xdeadbeef),
-	})
+	rnd             = &lockedSource{src: rand.NewSource(0xdeadbeef)}
```
Comment (coocood) [inline leveldb/memdb/memdb.go]: it is a source, not random instance.

**Comment by tiancaiamao** [comment]
PTAL @coocood @shenli

**Comment by coocood** [comment]
LGTM

File: leveldb/memdb/memdb.go [hunk 2]
```diff
@@ -219,9 +239,8 @@ type DB struct {
 }
 
 func (p *DB) randHeight() (h int) {
-	const branching = 4
 	h = 1
-	for h < tMaxHeight && p.rnd.Int()%branching == 0 {
+	for h < tMaxHeight && p.rnd.bitN(2) == 0 {
```
Comment (zimulala) [inline leveldb/memdb/memdb.go]: Can we set '2' to a constant and add some comments?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**